File: lib/expression/pm.js
// Helpers for working with the `pm` (plus-minus) operator.
//
// AST shape: `pm` is unary, analogous to unary `-`. A pm-bearing expression
// like `5 \pm 3` is represented as `["+", 5, ["pm", 3]]`. Each `["pm", x]`
// denotes the set `{x, -x}` with an independent sign choice.

// Maximum number of `pm` operators allowed in a single expression for
// sign-expansion. `expand_pm_signs` produces 2^MAX_PM_COUNT variants, so
// raising this trades exponential work for the ability to handle more
// independent ± choices.
const MAX_PM_COUNT = 10;
const MAX_PM_EXPANSIONS = 1 << MAX_PM_COUNT; // 1024

// Returns true if `tree` contains any `pm` operator anywhere in its subtree.
export function contains_pm(tree) {
  if (!Array.isArray(tree)) return false;
  if (tree[0] === "pm") return true;
  for (let i = 1; i < tree.length; i++) {
    if (contains_pm(tree[i])) return true;
  }
  return false;
}

// Returns the number of `pm` operators anywhere in `tree`.
export function count_pm(tree) {
  if (!Array.isArray(tree)) return 0;
  let n = tree[0] === "pm" ? 1 : 0;
  for (let i = 1; i < tree.length; i++) {
    n += count_pm(tree[i]);
  }
  return n;
}

// Enumerate all 2^n sign assignments for the `pm` operators in `tree`.
// Each `["pm", x]` is replaced either by `x` (sign = +) or by `["-", x]`
// (sign = -). Throws if the count would exceed MAX_PM_COUNT.
export function expand_pm_signs(tree) {
  const n = count_pm(tree);
  if (n === 0) return [tree];
  if (n > MAX_PM_COUNT) {
    throw new Error(
      `pm: cannot expand ${n} plus-minus operators (limit is ${MAX_PM_COUNT} → ${MAX_PM_EXPANSIONS} combinations)`,
    );
  }
  const total = 1 << n;
  const results = [];
  for (let mask = 0; mask < total; mask++) {
    const counter = { idx: 0 };
    results.push(replace_pm(tree, mask, counter));
  }
  return results;
}

function replace_pm(tree, mask, counter) {
  if (!Array.isArray(tree)) return tree;
  if (tree[0] === "pm") {
    const bit = (mask >> counter.idx) & 1;
    counter.idx += 1;
    const inner = replace_pm(tree[1], mask, counter);
    return bit === 0 ? inner : ["-", inner];
  }
  const out = [tree[0]];
  for (let i = 1; i < tree.length; i++) {
    out.push(replace_pm(tree[i], mask, counter));
  }
  return out;
}

File: packages/math-expressions-rs/src/pm.rs
//! Plus-minus (`±`) operator helpers — port of JS `lib/expression/pm.js`.
//!
//! `pm` is unary, analogous to unary `-`: `5 ± 3` is `Add(5, OtherOp("pm",[3]))`
//! and each `["pm", x]` denotes the two-element set `{x, -x}` with an
//! independent sign choice. These helpers detect, count, and sign-expand `pm`
//! nodes; numerical set-equality of `pm`-bearing expressions lives in
//! `equality::pm_equals` (the consumer, mirroring JS `equality/pm-numerical.js`).
//!
//! **Supported containers.** `±` may appear in scalars, relations (equations and
//! inequalities), and sequences — tuples, vectors, altvectors, arrays, lists,
//! and sets — which `equality::pm_equals` compares componentwise. **`±` is not
//! supported inside a matrix entry or an interval endpoint**: those are compared
//! only as opaque wholes, so a `pm`-bearing matrix or interval is equal only to
//! a structurally identical one (`equals` cannot see through the `±` there).
//! Callers should not place `±` inside matrices or intervals.

use crate::expr::Expr;
use crate::norm::syntactic::map_children;

/// Maximum number of `pm` operators allowed for sign-expansion.
/// `expand_pm_signs` produces `2^n` variants, so this caps the work at 1024.
pub const MAX_PM_COUNT: u32 = 10;

/// `expand_pm_signs` refused to expand because the `pm` count exceeds
/// [`MAX_PM_COUNT`] (`2^n` would be too many variants).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct PmOverflow {
    pub count: usize,
}

impl std::fmt::Display for PmOverflow {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "pm: cannot expand {} plus-minus operators (limit is {})",
            self.count, MAX_PM_COUNT
        )
    }
}

/// Is `e` a single `±x` node (`OtherOp("pm", [x])`)?
pub(crate) fn is_pm(e: &Expr) -> bool {
    matches!(e, Expr::OtherOp(name, args) if name.name() == "pm" && args.len() == 1)
}

/// Build a `±inner` node.
pub(crate) fn make_pm(inner: Expr) -> Expr {
    crate::parse::common::other_op("pm", vec![inner])
}

/// Does `e` contain any `pm` operator anywhere in its subtree?
pub fn contains_pm(e: &Expr) -> bool {
    is_pm(e) || e.children().iter().any(|c| contains_pm(c))
}

/// Number of `pm` operators anywhere in `e`.
pub fn count_pm(e: &Expr) -> usize {
    let here = usize::from(is_pm(e));
    here + e.children().iter().map(|c| count_pm(c)).sum::<usize>()
}

/// Enumerate all `2^n` sign assignments for the `pm` operators in `e`. Each
/// `["pm", x]` becomes either `x` (sign `+`) or `Neg(x)` (sign `−`). Returns a
/// single-element `[e]` when there are no `pm` nodes, or [`PmOverflow`] when the
/// count exceeds [`MAX_PM_COUNT`].
pub fn expand_pm_signs(e: &Expr) -> Result<Vec<Expr>, PmOverflow> {
    let n = count_pm(e);
    if n == 0 {
        return Ok(vec![e.clone()]);
    }
    if n > MAX_PM_COUNT as usize {
        return Err(PmOverflow { count: n });
    }
    let total = 1usize << n;
    let mut out = Vec::with_capacity(total);
    for mask in 0..total {
        let mut idx = 0;
        out.push(replace_pm(e, mask, &mut idx));
    }
    Ok(out)
}

/// Replace each `pm` node by its `+` or `−` variant per the corresponding bit of
/// `mask`. `idx` walks the `pm` nodes in the same pre-order as [`count_pm`], so
/// bit `k` of `mask` controls the `k`-th `pm`.
fn replace_pm(e: &Expr, mask: usize, idx: &mut usize) -> Expr {
    if is_pm(e) {
        let Expr::OtherOp(_, args) = e else { unreachable!() };
        let bit = (mask >> *idx) & 1;
        *idx += 1;
        let inner = replace_pm(&args[0], mask, idx);
        return if bit == 0 {
            inner
        } else {
            Expr::Neg(Box::new(inner))
        };
    }
    map_children(e, |c| replace_pm(c, mask, idx))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{Expr, TextToAst, TextToAstOptions};

    fn parse(s: &str) -> Expr {
        TextToAst::new(TextToAstOptions::default()).convert(s).unwrap()
    }

    #[test]
    fn detects_and_counts() {
        assert!(!contains_pm(&parse("5 + 3")));
        assert!(contains_pm(&parse("5 ± 3")));
        assert_eq!(count_pm(&parse("5 + 3")), 0);
        assert_eq!(count_pm(&parse("5 ± 3")), 1);
        assert_eq!(count_pm(&parse("5 ± 3 ± 4")), 2);
    }

    #[test]
    fn expands_2_to_the_n() {
        assert_eq!(expand_pm_signs(&parse("5")).unwrap().len(), 1);
        assert_eq!(expand_pm_signs(&parse("5 ± 3")).unwrap().len(), 2);
        assert_eq!(expand_pm_signs(&parse("5 ± 3 ± 4")).unwrap().len(), 4);
    }

    #[test]
    fn overflow_beyond_limit() {
        let many = "x ".to_string() + &"± 1 ".repeat((MAX_PM_COUNT + 1) as usize);
        let e = parse(many.trim());
        assert_eq!(count_pm(&e), (MAX_PM_COUNT + 1) as usize);
        assert!(expand_pm_signs(&e).is_err());
    }
}
